c7n/filters/iamaccess.py

```diff
@@ -123,11 +124,12 @@ def handle_principal(self, s):
             return True
         # Skip service principals
         if 'Service' in s['Principal']:
-            s['Principal'].pop('Service')
```

Comment:
I'm interested in the rare issue you mentioned that can lead to a null principal here. It might be helpful to try to reproduce/capture that (ideally in a failing test) so we can make sure we solve for it here. I like that your change tries to bail out on Service-only principals without leaving side effects though :+1: 

Stepping up a level it sounds like we're doing a few levels of checks and transforms to:
* Handle simple string principals
* Check for `AWS` or `Federated` principals
* Skip `Service` principals

I'm wondering if we could handle all of those together by building a single flat list of `AWS` and `Federated` principals and looping over that below. That would seem to also help address your point about hitting the "too many principals" problem unexpectedly.

Reply:
Thanks, @ajkerrigan.  I haven’t had an opportunity yet to make the additional changes that you suggested. Should I split out the two fixes so that at least one can be merged now and the other later when I am able to make the suggested changes and/or supply a negative test case to reproduce the failure?

Reply:
The work you put in already is much appreciated. I added some failing checks to the test you wrote, based on following bread crumbs from the information in #7153 . Then tweaked the logic a bit to handle cases with different combinations of principal types. There may still be wrinkles to iron out, but hopefully it's another step forward. Thanks again.